pkg/types/builder.go

```diff
@@ -39,28 +39,38 @@ const (
 // NewBuilder returns a new Builder.
 func NewBuilder(pkg *types.Package) *Builder {
 	return &Builder{
-		Package:  pkg,
-		comments: twtypes.Comments{},
+		Package: pkg,
+		Generated: Generated{
+			Comments: twtypes.Comments{},
+		},
 	}
 }
 
+// Generated stores the information of the generated types
+type Generated struct {
+	Types    []*types.Named
+	Comments twtypes.Comments
+
+	ForProviderType *types.Named
+	AtProviderType  *types.Named
+}
+
 // Builder is used to generate Go type equivalence of given Terraform schema.
 type Builder struct {
 	Package *types.Package
-
-	genTypes []*types.Named
-	comments twtypes.Comments
+	Generated
 }
 
-// Build returns parameters and observation types built out of Terraform schema.
-func (g *Builder) Build(cfg *config.Resource) ([]*types.Named, twtypes.Comments, error) {
-	_, _, err := g.buildResource(cfg.TerraformResource, cfg, nil, nil, cfg.Kind)
-	return g.genTypes, g.comments, errors.Wrapf(err, "cannot build the types")
+// Build returns parameters and observation Types built out of Terraform schema.
+func (g *Builder) Build(cfg *config.Resource) (Generated, error) {
+	var err error
+	g.ForProviderType, g.AtProviderType, err = g.buildResource(cfg.TerraformResource, cfg, nil, nil, cfg.Kind)
+	return g.Generated, errors.Wrapf(err, "cannot build the Types")
 }
 
 func (g *Builder) buildResource(res *schema.Resource, cfg *config.Resource, tfPath []string, xpPath []string, names ...string) (*types.Named, *types.Named, error) { //nolint:gocyclo
 	// NOTE(muvaf): There can be fields in the same CRD with same name but in
-	// different types. Since we generate the type using the field name, there
+	// different Types. Since we generate the type using the field name, there
```

Comment:
Seems like there are quite a few unintended changes from `types` to `Types` done by IDE.